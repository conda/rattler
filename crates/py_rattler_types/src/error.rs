use std::error::Error;

use pyo3::{create_exception, exceptions::PyException, PyErr};
use rattler_conda_types::{ParseArchError, ParsePlatformError};
use thiserror::Error;

#[derive(Error, Debug)]
#[allow(clippy::enum_variant_names)]
pub enum PyRattlerError {
    #[error(transparent)]
    ParsePlatformError(#[from] ParsePlatformError),
    #[error(transparent)]
    ParseArchError(#[from] ParseArchError),
}

fn pretty_print_error(mut err: &dyn Error) -> String {
    let mut result = err.to_string();
    while let Some(source) = err.source() {
        result.push_str(&format!("\nCaused by: {source}"));
        err = source;
    }
    result
}

impl From<PyRattlerError> for PyErr {
    fn from(value: PyRattlerError) -> Self {
        match value {
            PyRattlerError::ParsePlatformError(err) => {
                ParsePlatformException::new_err(pretty_print_error(&err))
            }
            PyRattlerError::ParseArchError(err) => {
                ParseArchException::new_err(pretty_print_error(&err))
            }
        }
    }
}

create_exception!(exceptions, ParsePlatformException, PyException);
create_exception!(exceptions, ParseArchException, PyException);
