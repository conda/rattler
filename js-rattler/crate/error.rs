use rattler_conda_types::version_spec::ParseVersionSpecError;
use rattler_conda_types::{ParseVersionError, VersionBumpError, VersionExtendError};
use thiserror::Error;
use wasm_bindgen::JsValue;

#[derive(Debug, Error)]
pub enum JsError {
    #[error(transparent)]
    InvalidVersion(#[from] ParseVersionError),
    #[error(transparent)]
    VersionExtendError(#[from] VersionExtendError),
    #[error(transparent)]
    VersionBumpError(#[from] VersionBumpError),

    #[error(transparent)]
    ParseVersionSpecError(#[from] ParseVersionSpecError),
}

pub type JsResult<T> = Result<T, JsError>;

impl From<JsError> for JsValue {
    fn from(error: JsError) -> Self {
        match error {
            JsError::InvalidVersion(error) => JsValue::from_str(&error.to_string()),
            JsError::VersionExtendError(error) => JsValue::from_str(&error.to_string()),
            JsError::VersionBumpError(error) => JsValue::from_str(&error.to_string()),
            JsError::ParseVersionSpecError(error) => JsValue::from_str(&error.to_string()),
        }
    }
}
