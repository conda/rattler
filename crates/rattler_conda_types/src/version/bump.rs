use std::borrow::Cow;

use thiserror::Error;

use crate::{Component, Version};

use super::{segment::Segment, ComponentVec, SegmentVec};

/// `VersionBumpType` is used to specify the type of bump to perform on a version.
#[derive(Clone)]
pub enum VersionBumpType {
    /// Bump the major version number.
    Major,
    /// Bump the minor version number.
    Minor,
    /// Bump the patch version number.
    Patch,
    /// Bump the last version number.
    Last,
    /// Bump a given segment. If negative, count from the end.
    Segment(i32),
}

/// `VersionBumpError` is used to specify the type of error that occurred when bumping a version.
#[derive(Error, Debug, PartialEq)]
pub enum VersionBumpError {
    /// Invalid segment index.
    #[error("cannot bump the segment '{index:?}' of a version if it's not present")]
    InvalidSegment {
        /// The segment index that was attempted to be bumped.
        index: i32,
    },

    /// Could not extend the version
    #[error("could not extend the version: {0}")]
    VersionExtendError(#[from] crate::VersionExtendError),
}

impl Version {
    /// Add alpha specifier to the end of the version when the last element does not contain an `iden` component.
    ///
    /// For example, `1.0.0` will become `1.0.0.0a0`.
    /// If the last version element contains a character, it's not modified (e.g. `1.0.0a` will remain `1.0.0a`).
    pub fn with_alpha(&self) -> Cow<'_, Self> {
        let last_segment = self.segments().last().expect("at least one segment");
        // check if there is an iden component in the last segment
        let has_iden = last_segment.components().any(|c| c.as_iden().is_some());
        if has_iden {
            return Cow::Borrowed(self);
        }
        let local_segment_index = self.local_segment_index().unwrap_or(self.segments.len());
        let mut segments = self.segments[0..local_segment_index].to_vec();
        let components_offset = segments.iter().map(|s| s.len() as usize).sum::<usize>()
            + usize::from(self.has_epoch());

        segments.push(Segment::new(3).unwrap().with_separator(Some('.')).unwrap());
        segments.extend(self.segments[local_segment_index..].iter());

        let mut components = self.components.clone();
        components.insert(components_offset, Component::Numeral(0));
        components.insert(components_offset + 1, Component::Iden("a".into()));
        components.insert(components_offset + 2, Component::Numeral(0));

        let flags = if let Some(local_segment_index) = self.local_segment_index() {
            self.flags
                .with_local_segment_index((local_segment_index + 1) as u8)
                .unwrap()
        } else {
            self.flags
        };

        Cow::Owned(Version {
            components,
            segments: segments.into(),
            flags,
        })
    }

    /// Remove the local segment from the version if it exists.
    /// Returns a new version without the local segment.
    ///
    /// For example, `1.0.0+3.4` will become `1.0.0`.
    pub fn remove_local(&self) -> Cow<'_, Self> {
        if let Some(local_segment_index) = self.local_segment_index() {
            let segments = self.segments[0..local_segment_index].to_vec();
            let components_offset = segments.iter().map(|s| s.len() as usize).sum::<usize>()
                + usize::from(self.has_epoch());
            let mut components = self.components.clone();
            components.drain(components_offset..);

            Cow::Owned(Version {
                components,
                segments: segments.into(),
                flags: self.flags.with_local_segment_index(0).unwrap(),
            })
        } else {
            return Cow::Borrowed(self);
        }
    }

    /// Returns a new version after bumping it according to the specified bump type.
    /// Note: if a version ends with a character, the next bigger version will use `a` as the character.
    /// For example: `1.1l` -> `1.2a`, but also `1.1.0alpha` -> `1.1.1a`.
    pub fn bump(&self, bump_type: VersionBumpType) -> Result<Self, VersionBumpError> {
        // Sanity check whether the version has enough segments for this bump type.
        let segment_count = self.segment_count();
        let segment_to_bump = match bump_type {
            VersionBumpType::Major => 0,
            VersionBumpType::Minor => 1,
            VersionBumpType::Patch => 2,
            VersionBumpType::Last => {
                if segment_count > 0 {
                    segment_count - 1
                } else {
                    0
                }
            }
            VersionBumpType::Segment(index_to_bump) => {
                let computed_index = if index_to_bump < 0 {
                    index_to_bump + segment_count as i32
                } else {
                    index_to_bump
                };
                if computed_index < 0 {
                    return Err(VersionBumpError::InvalidSegment {
                        index: index_to_bump,
                    });
                }
                computed_index as usize
            }
        };

        // Add the necessary segments to the version if it's too short.
        let version = self.extend_to_length(segment_to_bump + 1)?;

        let mut components = ComponentVec::new();
        let mut segments = SegmentVec::new();
        let mut flags = version.flags;

        // Copy the optional epoch.
        if let Some(epoch) = version.epoch_opt() {
            components.push(Component::Numeral(epoch));
            flags = flags.with_has_epoch(true);
        }

        // Copy over all the segments and bump the last segment.
        for (idx, segment_iter) in version.segments().enumerate() {
            let segment = segment_iter.segment;
            let mut segment_components =
                segment_iter.components().cloned().collect::<ComponentVec>();

            // Handle semantic versioning resets
            match bump_type {
                VersionBumpType::Major => {
                    if idx == 0 {
                        // Increment major version
                        if let Some(num) = segment_components
                            .iter_mut()
                            .find_map(Component::as_number_mut)
                        {
                            *num += 1;
                        }
                        // Change identifier to 'a' if present
                        if let Some(iden) = segment_components
                            .iter_mut()
                            .filter_map(Component::as_iden_mut)
                            .next_back()
                        {
                            *iden = "a".into();
                        }
                    } else if idx <= 2 {
                        // Reset minor and patch to 0
                        if let Some(num) = segment_components
                            .iter_mut()
                            .find_map(Component::as_number_mut)
                        {
                            *num = 0;
                        }
                    }
                }
                VersionBumpType::Minor => {
                    if idx == 1 {
                        // Increment minor version
                        if let Some(num) = segment_components
                            .iter_mut()
                            .find_map(Component::as_number_mut)
                        {
                            *num += 1;
                        }
                    } else if idx == 2 {
                        // Reset patch to 0
                        if let Some(num) = segment_components
                            .iter_mut()
                            .find_map(Component::as_number_mut)
                        {
                            *num = 0;
                        }
                    }
                }
                VersionBumpType::Patch => {
                    if idx == 2 {
                        // Increment patch version
                        if let Some(num) = segment_components
                            .iter_mut()
                            .find_map(Component::as_number_mut)
                        {
                            *num += 1;
                        }
                        // Change identifier to 'a' if present
                        if let Some(iden) = segment_components
                            .iter_mut()
                            .filter_map(Component::as_iden_mut)
                            .next_back()
                        {
                            *iden = "a".into();
                        }
                    }
                }
                _ => {
                    // For Last and Segment types, keep original behavior
                    if segment_to_bump == idx {
                        let last_numeral_component = segment_components
                            .iter_mut()
                            .filter_map(Component::as_number_mut)
                            .next_back()
                            .expect(
                                "every segment must at least contain a single numeric component",
                            );
                        *last_numeral_component += 1;

                        // If the segment ends with an ascii character, make it `a` instead of whatever it says
                        let last_iden_component = segment_components
                            .iter_mut()
                            .filter_map(Component::as_iden_mut)
                            .next_back();

                        if let Some(last_iden_component) = last_iden_component {
                            *last_iden_component = "a".into();
                        }
                    }
                }
            }

            let has_implicit_default =
                segment.has_implicit_default() && segment_components[0] == Component::default();
            let start_idx = usize::from(has_implicit_default);

            let component_count = segment_components.len();
            for component in segment_components.into_iter().skip(start_idx) {
                components.push(component);
            }

            let segment = Segment::new((component_count - start_idx) as _)
                .expect("there will be no more components than in the previous segment")
                .with_implicit_default(has_implicit_default)
                .with_separator(segment.separator())
                .expect("copying the segment should just work");

            segments.push(segment);
        }

        if version.has_local() {
            let segment_idx = segments.len() as u8;
            for segment_iter in self.local_segments() {
                for component in segment_iter.components().cloned() {
                    components.push(component);
                }
                segments.push(segment_iter.segment);
            }
            flags = flags
                .with_local_segment_index(segment_idx)
                .expect("this should never fail because no new segments are added");
        }

        Ok(Self {
            components,
            segments,
            flags,
        })
    }
}

#[cfg(test)]
mod test {
    use crate::{Version, VersionBumpType};
    use rstest::rstest;
    use std::str::FromStr;

    // Original conda-style tests
    #[rstest]
    #[case("1.1", "1.2")]
    #[case("1.1l", "1.2a")]
    #[case("5!1.alpha+3.4", "5!1.1a+3.4")]
    fn bump_last(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Last)
                .unwrap(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case("1.1", "2.0")]
    #[case("2.1l", "3.0l")]
    #[case("5!1.alpha+3.4", "5!2.alpha+3.4")]
    #[case("1.2.3", "2.0.0")]
    #[case("0.1.0", "1.0.0")]
    #[case("1.2.3-alpha", "2.0.0-alpha")]
    #[case("9d", "10a")]
    fn bump_major(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Major)
                .unwrap(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case("1.1", "1.2")]
    #[case("2.1l", "2.2l")]
    #[case("5!1.alpha+3.4", "5!1.1alpha+3.4")]
    #[case("1.2.3", "1.3.0")]
    #[case("1.9.0", "1.10.0")]
    #[case("0.1.0", "0.2.0")]
    fn bump_minor(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Minor)
                .unwrap(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case("1.1.9", "1.1.10")]
    #[case("2.1l.5alpha", "2.1l.6a")]
    #[case("5!1.8.alpha+3.4", "5!1.8.1a+3.4")]
    #[case("1.2.3", "1.2.4")]
    #[case("1.2.9", "1.2.10")]
    #[case("0.1.0", "0.1.1")]
    #[case("1.1.1e", "1.1.2a")]
    fn bump_patch(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Patch)
                .unwrap(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case(0, "1.1.9", "2.1.9")]
    #[case(1, "1.1.9", "1.2.9")]
    #[case(2, "1.1.9", "1.1.10")]
    #[case(-1, "1.1.9", "1.1.10")]
    #[case(-2, "1.1.9", "1.2.9")]
    #[case(-3, "1.1.9", "2.1.9")]
    #[case(0, "9d", "10a")]
    #[case(5, "1.2.3", "1.2.3.0.0.1")]
    #[case(2, "9d", "9d.0.1")]
    fn bump_segment(#[case] idx: i32, #[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Segment(idx))
                .unwrap(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case(0, "1.1.9", "2.1.9.0a0")]
    #[case(2, "1.0.0", "1.0.1.0a0")]
    #[case(2, "1.0.0a", "1.0.1a")]
    #[case(2, "1.0.0f", "1.0.1a")]
    #[case(2, "5!1.0.0", "5!1.0.1.0a0")]
    #[case(2, "5!1.0.0+3.4", "5!1.0.1.0a0+3.4")]
    fn with_alpha(#[case] idx: i32, #[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .bump(VersionBumpType::Segment(idx))
                .unwrap()
                .with_alpha()
                .into_owned(),
            Version::from_str(expected).unwrap()
        );
    }

    #[rstest]
    #[case("1.1.9", "1.1.9")]
    #[case("1.0.0+3", "1.0.0")]
    #[case("1.0.0+3.4", "1.0.0")]
    #[case("1.0.0+3.4alpha.2.4", "1.0.0")]
    #[case("5!1.0.0+3.4alpha.2.4", "5!1.0.0")]
    fn remove_local(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(
            Version::from_str(input)
                .unwrap()
                .remove_local()
                .into_owned(),
            Version::from_str(expected).unwrap()
        );
    }
}
