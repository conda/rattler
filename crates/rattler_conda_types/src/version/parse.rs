use super::{Component, StrictVersion, Version};
use crate::version::flags::Flags;
use crate::version::segment::Segment;
use crate::version::{ComponentVec, SegmentVec};
use nom::branch::alt;
use nom::bytes::complete::tag_no_case;
use nom::character::complete::{alpha1, char, digit1, one_of};
use nom::combinator::{map, opt, value};
use nom::error::{ErrorKind, FromExternalError, ParseError};
use nom::sequence::terminated;
use nom::IResult;
use smallvec::SmallVec;
use std::{
    convert::Into,
    default::Default,
    error::Error,
    fmt::{Display, Formatter},
    num::ParseIntError,
    result::Result,
    str::FromStr,
};
use thiserror::Error;

/// An error that occurred during parsing of a string to a version.
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ParseVersionError {
    /// The original string that was the input of the parser
    pub version: String,

    /// The type of parse error that occurred
    pub kind: ParseVersionErrorKind,
}

impl Display for ParseVersionError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "malformed version string '{}': {}",
            &self.version, &self.kind
        )
    }
}

impl Error for ParseVersionError {}

impl ParseVersionError {
    /// Create a new parse error
    pub fn new(text: impl Into<String>, kind: ParseVersionErrorKind) -> Self {
        Self {
            version: text.into(),
            kind,
        }
    }
}

/// The type of parse error that occurred when parsing a version string.
#[derive(Debug, Eq, PartialEq, Clone, Error)]
pub enum ParseVersionErrorKind {
    /// The string was empty
    #[error("empty string")]
    Empty,
    /// The epoch was not an integer value
    #[error("epoch is not a number")]
    EpochMustBeInteger(ParseIntError),
    /// The string contained an invalid numeral
    #[error("invalid number")]
    InvalidNumeral(ParseIntError),
    /// The string contained an empty version component
    #[error("expected a version component e.g. `2` or `rc`")]
    EmptyVersionComponent,
    /// Too many segments.
    #[error("the version string contains too many version segments")]
    TooManySegments,
    /// Too many segments.
    #[error("there are too many components in a single segment")]
    TooManyComponentsInASegment,
    /// Expected a version component
    #[error("expected a version component e.g. `2` or `rc`")]
    ExpectedComponent,
    /// Expected a segment separator
    #[error("expected a '.', '-', or '_'")]
    ExpectedSegmentSeparator,
    /// Cannot mix and match dashes and underscores
    #[error("cannot use both underscores and dashes as version segment separators")]
    CannotMixAndMatchDashesAndUnderscores,
    /// Expected the end of the string
    #[error("encountered more characters but expected none")]
    ExpectedEof,
    /// Nom error
    #[error("{0:?}")]
    Nom(ErrorKind),
}

impl<'i> ParseError<&'i str> for ParseVersionErrorKind {
    fn from_error_kind(_: &'i str, kind: ErrorKind) -> Self {
        ParseVersionErrorKind::Nom(kind)
    }

    fn append(_: &'i str, _: ErrorKind, other: Self) -> Self {
        other
    }
}

impl<'i> FromExternalError<&'i str, ParseVersionErrorKind> for ParseVersionErrorKind {
    fn from_external_error(_: &'i str, _: ErrorKind, e: ParseVersionErrorKind) -> Self {
        e
    }
}

/// Parses the epoch part of a version. This is a number followed by `'!'` at the start of the
/// version string.
pub fn epoch_parser(input: &str) -> IResult<&str, u64, ParseVersionErrorKind> {
    let (rest, digits) = terminated(digit1, char('!'))(input)?;
    let epoch = digits
        .parse()
        .map_err(ParseVersionErrorKind::EpochMustBeInteger)
        .map_err(nom::Err::Failure)?;
    Ok((rest, epoch))
}

/// Parses a numeral from the input, fails if the parsed digits cannot be represented by an `u64`.
fn numeral_parser(input: &str) -> IResult<&str, u64, ParseVersionErrorKind> {
    let (rest, digits) = digit1(input)?;
    match u64::from_str(digits) {
        Ok(numeral) => Ok((rest, numeral)),
        Err(e) => Err(nom::Err::Failure(ParseVersionErrorKind::InvalidNumeral(e))),
    }
}

/// Parses a single version [`Component`].
fn component_parser<'i>(input: &'i str) -> IResult<&'i str, Component, ParseVersionErrorKind> {
    alt((
        // Parse a numeral
        map(numeral_parser, Component::Numeral),
        // Parse special case components
        value(Component::Post, tag_no_case("post")),
        value(Component::Dev, tag_no_case("dev")),
        // Parse an identifier
        map(alpha1, |alpha: &'i str| {
            Component::Iden(alpha.to_lowercase().into_boxed_str())
        }),
    ))(input)
}

/// Parses a version segment from a list of components.
fn segment_parser<'i>(
    components: &mut ComponentVec,
    input: &'i str,
) -> IResult<&'i str, Segment, ParseVersionErrorKind> {
    // Parse the first component of the segment
    let (mut rest, first_component) = match component_parser(input) {
        Ok(result) => result,
        // Convert undefined parse errors into an expect error
        Err(nom::Err::Error(ParseVersionErrorKind::Nom(_))) => {
            return Err(nom::Err::Error(ParseVersionErrorKind::ExpectedComponent))
        }
        Err(e) => return Err(e),
    };

    // If the first component is not numeric we add a default component since each segment must
    // always start with a number.
    let mut component_count = 0u16;
    let has_implicit_default = !first_component.is_numeric();

    // Add the first component
    components.push(first_component);
    component_count += 1;

    // Loop until we can't find any more components
    loop {
        let (remaining, component) = match opt(component_parser)(rest) {
            Ok((i, o)) => (i, o),
            Err(e) => {
                // Remove any components that we may have added.
                components.drain(components.len() - (component_count as usize)..);
                return Err(e);
            }
        };
        if let Some(component) = component {
            components.push(component);
            component_count = match component_count.checked_add(1) {
                Some(length) => length,
                None => {
                    return Err(nom::Err::Failure(
                        ParseVersionErrorKind::TooManyComponentsInASegment,
                    ))
                }
            }
        } else {
            let segment = Segment::new(component_count)
                .ok_or(nom::Err::Failure(
                    ParseVersionErrorKind::TooManyComponentsInASegment,
                ))?
                .with_implicit_default(has_implicit_default);

            break Ok((remaining, segment));
        }
        rest = remaining;
    }
}

/// Parses a trailing underscore or dash.
fn trailing_dash_underscore_parser(
    input: &str,
    dash_or_underscore: Option<char>,
) -> IResult<&str, (Option<Component>, Option<char>), ParseVersionErrorKind> {
    // Parse a - or _. Return early if it cannot be found.
    let (rest, Some(separator)) = opt(one_of::<_, _, (&str, ErrorKind)>("-_"))(input)
        .map_err(|e| e.map(|(_, kind)| ParseVersionErrorKind::Nom(kind)))?
    else {
        return Ok((input, (None, dash_or_underscore)));
    };

    // Make sure dashes and underscores are not mixed.
    let dash_or_underscore = match (dash_or_underscore, separator) {
        (None, '-') => Some('-'),
        (None, '_') => Some('_'),
        (Some('-'), '_') | (Some('_'), '-') => {
            return Err(nom::Err::Error(
                ParseVersionErrorKind::CannotMixAndMatchDashesAndUnderscores,
            ))
        }
        _ => dash_or_underscore,
    };

    Ok((
        rest,
        (
            Some(Component::UnderscoreOrDash {
                is_dash: separator == '-',
            }),
            dash_or_underscore,
        ),
    ))
}

fn version_part_parser<'i>(
    components: &mut ComponentVec,
    segments: &mut SegmentVec,
    input: &'i str,
    dash_or_underscore: Option<char>,
) -> IResult<&'i str, Option<char>, ParseVersionErrorKind> {
    let mut dash_or_underscore = dash_or_underscore;
    let mut recovery_segment_idx = segments.len();

    // Parse the first segment of the version. It must exists.
    let (mut input, first_segment_length) = segment_parser(components, input)?;
    segments.push(first_segment_length);

    // Iterate over any additional segments that we find.
    let result = loop {
        // Parse a version segment separator.
        let (rest, separator) = match opt(one_of("-._"))(input) {
            Ok((_, None)) => {
                // No additional separator found, exit early.
                return Ok((input, dash_or_underscore));
            }
            Ok((rest, Some(separator))) => (rest, separator),

            Err(nom::Err::Error(_)) => {
                // If an error occurred we convert it to a segment separator not found error instead.
                break Err(nom::Err::Error(
                    ParseVersionErrorKind::ExpectedSegmentSeparator,
                ));
            }

            // Failure are propagated
            Err(e) => break Err(e),
        };

        // Make sure dashes and underscores are not mixed.
        match (dash_or_underscore, separator) {
            (None, '-') => dash_or_underscore = Some('-'),
            (None, '_') => dash_or_underscore = Some('_'),
            (Some('-'), '_') | (Some('_'), '-') => {
                break Err(nom::Err::Failure(
                    ParseVersionErrorKind::CannotMixAndMatchDashesAndUnderscores,
                ))
            }
            _ => {}
        }

        // Parse the next segment.
        let (rest, segment) = match segment_parser(components, rest) {
            Ok(result) => result,
            Err(nom::Err::Error(_)) => {
                // If parsing of a segment failed, check if perhaps the separator is followed by an
                // underscore or dash.
                match trailing_dash_underscore_parser(rest, dash_or_underscore)? {
                    (rest, (Some(component), dash_or_underscore)) => {
                        // We are parsing multiple dashes or underscores ("..__"), add a new segment
                        // just for the trailing underscore/dash
                        components.push(component);
                        segments.push(
                            Segment::new(1)
                                .unwrap()
                                .with_implicit_default(true)
                                .with_separator(Some(separator))
                                .unwrap(),
                        );

                        // Since the trailing is always at the end we immediately return
                        return Ok((rest, dash_or_underscore));
                    }
                    (rest, (None, dash_or_underscore)) if separator == '-' || separator == '_' => {
                        // We are parsing a single dash or underscore (".._"), update the last
                        // segment we added
                        let segment = segments
                            .last_mut()
                            .expect("there must be at least one segment added");
                        components.push(Component::UnderscoreOrDash {
                            is_dash: separator == '-',
                        });

                        *segment = segment
                            .len()
                            .checked_add(1)
                            .and_then(|len| segment.with_component_count(len))
                            .ok_or(nom::Err::Failure(
                                ParseVersionErrorKind::TooManyComponentsInASegment,
                            ))?;

                        // Since the trailing is always at the end we immediately return
                        return Ok((rest, dash_or_underscore));
                    }
                    _ => return Ok((input, dash_or_underscore)),
                }
            }

            // Failures are propagated
            Err(e) => break Err(e),
        };

        segments.push(
            segment
                .with_separator(Some(separator))
                .expect("unrecognized separator"),
        );
        recovery_segment_idx += 1;
        input = rest;
    };

    match result {
        // If there was an error, revert the `segment_lengths` array.
        Err(e) => {
            segments.drain(recovery_segment_idx..);
            Err(e)
        }
        Ok(separator) => Ok((input, separator)),
    }
}

pub fn version_parser(input: &str) -> IResult<&str, Version, ParseVersionErrorKind> {
    let mut components = SmallVec::default();
    let mut segments = SmallVec::default();
    let mut flags = Flags::default();

    // String must not be empty
    if input.is_empty() {
        return Err(nom::Err::Error(ParseVersionErrorKind::Empty));
    }

    // Parse an optional epoch.
    let (input, epoch) = opt(epoch_parser)(input)?;
    if let Some(epoch) = epoch {
        components.push(epoch.into());
        flags = flags.with_has_epoch(true);
    }

    // Parse the common part of the version
    let (rest, dash_or_underscore) =
        match version_part_parser(&mut components, &mut segments, input, None) {
            Ok(result) => result,
            Err(e) => return Err(e),
        };

    // Parse the local version part
    let rest = if let Ok((local_version_part, _)) = char::<_, (&str, ErrorKind)>('+')(rest) {
        let first_local_segment_idx = segments.len();

        // Encode the local segment index into the flags.
        match u8::try_from(first_local_segment_idx)
            .ok()
            .and_then(|idx| flags.with_local_segment_index(idx))
        {
            None => {
                // There are too many segments to be able to encode the local segment parts into the
                // special `flag` we store. The flags is 8 bits and the first bit is used to
                // indicate if there is an epoch or not. The remaining 7 bits are used to indicate
                // which segment is the first that belongs to the local version part. We can encode
                // at most 127 positions so if there are more segments in the common version part,
                // we cannot represent this version.
                return Err(nom::Err::Error(ParseVersionErrorKind::TooManySegments));
            }
            Some(updated_flags) => {
                flags = updated_flags;
            }
        }

        match version_part_parser(
            &mut components,
            &mut segments,
            local_version_part,
            dash_or_underscore,
        ) {
            Ok((rest, _)) => rest,
            Err(e) => return Err(e),
        }
    } else {
        rest
    };

    Ok((
        rest,
        Version {
            components,
            segments,
            flags,
        },
    ))
}

impl FromStr for Version {
    type Err = ParseVersionError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match version_parser(s) {
            Ok(("", version)) => Ok(version),
            Ok(_) => Err(ParseVersionError::new(
                s,
                ParseVersionErrorKind::ExpectedEof,
            )),
            Err(nom::Err::Failure(e) | nom::Err::Error(e)) => Err(ParseVersionError::new(s, e)),
            Err(_) => unreachable!("not streaming, so no other error possible"),
        }
    }
}

impl FromStr for StrictVersion {
    type Err = ParseVersionError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(StrictVersion(Version::from_str(s)?))
    }
}

#[cfg(test)]
mod test {
    use super::Version;
    use crate::version::parse::version_parser;
    use crate::version::SegmentFormatter;
    use serde::Serialize;
    use std::collections::BTreeMap;
    use std::path::Path;
    use std::str::FromStr;

    #[test]
    fn test_parse_star() {
        assert_eq!(
            version_parser("1.*"),
            Ok((".*", Version::from_str("1").unwrap()))
        );
    }

    #[test]
    fn test_parse() {
        #[derive(Debug, Serialize)]
        #[serde(untagged)]
        enum VersionOrError {
            Version(Version),
            Error(String),
        }

        let versions = [
            "$",
            ".",
            "1!1.2a.3-rc1",
            "1+",
            "1+$",
            "1+.",
            "1+2",
            "1-2-3",
            "1-2-3_",
            "1-2_3",
            "1.0.1_",
            "1.0.1-",
            "1.0.1post.za",
            "1@2",
            "1_",
            "1_2_3",
            "1_2_3_",
            "1__",
            "1___",
            "1--",
            "1-_",
            "1_-",
        ];

        let mut index_map: BTreeMap<String, VersionOrError> = BTreeMap::default();
        for version_str in versions {
            let version_or_error = match Version::from_str(version_str) {
                Ok(version) => {
                    assert_eq!(version_str, version.to_string().as_str());
                    VersionOrError::Version(version)
                }
                Err(e) => VersionOrError::Error(e.kind.to_string()),
            };
            index_map.insert(version_str.to_owned(), version_or_error);
        }

        insta::assert_debug_snapshot!(index_map);
    }

    /// Parse a large number of versions and see if parsing succeeded.
    /// TODO: This doesnt really verify that the parsing is correct. Maybe we can parse the version
    /// with Conda too and verify that the results match?
    #[test]
    fn test_parse_all() {
        let versions = std::fs::read_to_string(
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/parsed_versions.txt"),
        )
        .unwrap();
        for line in versions.lines() {
            // Skip comments and empty lines
            if line.trim_start().starts_with('#') || line.trim().is_empty() {
                continue;
            }

            let (version, debug_parsed) = line.split_once('=').unwrap();
            let parsed_version = Version::from_str(version).unwrap();
            let parsed_version_debug_string = format!(
                "{:?}",
                SegmentFormatter::new(
                    Some(parsed_version.epoch_opt().unwrap_or(0)),
                    parsed_version.segments()
                )
            );
            assert_eq!(parsed_version_debug_string, debug_parsed);
        }
    }
}
