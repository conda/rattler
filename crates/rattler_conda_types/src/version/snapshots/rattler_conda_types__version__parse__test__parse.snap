---
source: crates/rattler_conda_types/src/version/parse.rs
expression: index_map
---
{
    "$": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    ".": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1!1.2a.3-rc1": Version(
        Version {
            version: [[1], [1], [2, 'a'], [3], [0, 'rc', 1]],
            local: [],
        },
    ),
    "1+": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+$": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+.": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+2": Version(
        Version {
            version: [[0], [1]],
            local: [[2]],
        },
    ),
    "1--": Version(
        Version {
            version: [[0], [1], [0, '_']],
            local: [],
        },
    ),
    "1-2-3": Version(
        Version {
            version: [[0], [1], [2], [3]],
            local: [],
        },
    ),
    "1-2-3_": Error(
        "cannot use both underscores and dashes as version segment separators",
    ),
    "1-2_3": Error(
        "cannot use both underscores and dashes as version segment separators",
    ),
    "1-_": Error(
        "cannot use both underscores and dashes as version segment separators",
    ),
    "1.0.1-": Version(
        Version {
            version: [[0], [1], [0], [1, '_']],
            local: [],
        },
    ),
    "1.0.1_": Version(
        Version {
            version: [[0], [1], [0], [1, '_']],
            local: [],
        },
    ),
    "1.0.1post.za": Version(
        Version {
            version: [[0], [1], [0], [1, inf], [0, 'za']],
            local: [],
        },
    ),
    "1@2": Error(
        "encountered more characters but expected none",
    ),
    "1_": Version(
        Version {
            version: [[0], [1, '_']],
            local: [],
        },
    ),
    "1_-": Error(
        "cannot use both underscores and dashes as version segment separators",
    ),
    "1_2_3": Version(
        Version {
            version: [[0], [1], [2], [3]],
            local: [],
        },
    ),
    "1_2_3_": Version(
        Version {
            version: [[0], [1], [2], [3, '_']],
            local: [],
        },
    ),
    "1__": Version(
        Version {
            version: [[0], [1], [0, '_']],
            local: [],
        },
    ),
    "1___": Error(
        "encountered more characters but expected none",
    ),
}
